File: tests/net/test-udp.js
common = require('../common/common.js');

common.catch_errors(function(){

var dgram = require('dgram');
var assert = require('assert');

var child;

var kill = function() {
  if (child) {
    console.log('Killing children!');
    child.kill();
  }
};

switch (process.argv.slice(3,4).toString()) {
case "test-udp/UDPSend": {
	var received = 0;
	
	console.log("XXX: testing UDPSend");

	var socket = dgram.createSocket('udp4', function(buf, addr) { //'connection' listener
	  console.log("incoming datagram: " + buf);
	  if (++received === 2) {
	    socket.close();
	  }
	});

	socket.on('listening', function() { //'listening' listener
	  console.log('udp socket listening');
	  // start test
	  var child = common.runTest(process.argv[2],process.argv.slice(3));
	  
	  console.log("Running test: " + process.argv.slice(2).join(" "));

	});

  socket.bind(common.PORT);

	//process.on('uncaughtException', cleanup);
	process.on('exit', function(){
	 kill();
	 assert.equal(received,2);
	});

}	break;
case "test-udp/UDPReceive": {
	var sent = 0;
	console.log("XXX: testing UDPReceive");

	var socket = dgram.createSocket('udp4', function(buf, addr) { //'connection' listener
	  console.log("should not have received datagram! : " + buf);
	});

	var interval = setInterval(function(){
		var message = new Buffer("message" + sent);
		console.log("sending datagram: " + message);
		socket.send(message, 0, message.length, common.PORT, "127.0.0.1", function(err, bytes) {
			if ((sent = sent+1) >= 2) {
				clearInterval(interval);
				socket.close();
			}
		});
	}, 100);
	
  console.log('udp socket listening');
  // start test
  var child = common.runTest(process.argv[2],process.argv.slice(3));
  
  console.log("Running test: " + process.argv.slice(2).join(" "));
  
  socket.bind();

	//process.on('uncaughtException', cleanup);
	process.on('exit', function(){
	 kill();
	 assert.equal(sent,2);
	});
}	break;
default:
	console.log("XXX: testing ?");
}


timeout = setTimeout(function(){
	console.log("Test timeout!");
	kill();
	process.exit(-1);
}, 5000);


});

File: tests/net/test-udp.cc
#include "cctest.h"
#include "common.h"

#include "native.h"

using namespace native;

TEST(UDPSend) {
	int sent = 0;
	timer* interval = nullptr;
	udp* socket = nullptr;
	process::run([&](){
	  std::cout << "Running UDPSend" << std::endl << std::flush;

	  socket = udp::createSocket(UDP4,[](const Buffer& buf, std::shared_ptr<detail::net_addr> address){
	    CHECK(false == "should not have received datagram!");
	  });

	  socket->on<event::listening>([](){
	    std::cout << "socket listening" << std::endl << std::flush;
	  });

	  socket->on<event::error>([](const Exception& e){
	    std::cout << "error: " << e.message() << std::endl << std::flush;
	  });

	  socket->on<event::close>([](){
	    std::cout << "socket closed" << std::endl << std::flush;
	  });

	  socket->bind();

		interval = setInterval([&](){
		  Buffer buf ("message" + std::to_string(sent));
		  std::cout << "sending datagram: " << buf.str() << std::endl << std::flush;
		  try{
		    socket->send(buf,0,buf.size(),COMMON_PORT,"127.0.0.1",
		        [&](const detail::resval& r){
		          std::cout << "send completed with status: " << r.str() << std::endl << std::flush;
              if(++sent >= 2) {
                clearInterval(interval);
                socket->close();
              }
		        }
		    );
		  } catch(Exception* e) {
		    std::cerr << e->message() << std::endl << std::flush;
		  }
		}, 100);
	});

	CHECK(sent == 2);
}


TEST(UDPReceive) {
  int received = 0;
  timer* interval = nullptr;
  udp* socket = nullptr;
  process::run([&](){
    std::cout << "Running UDPReceive" << std::endl << std::flush;

    socket = udp::createSocket(UDP4,[&](const Buffer& buf, std::shared_ptr<detail::net_addr> address){
      std::cout << "received datagram: " << buf.str() << std::endl << std::flush;
      if(++received >= 2) {
        socket->close();
      }
    });

    socket->on<event::listening>([](){
      std::cout << "socket listening" << std::endl << std::flush;
    });

    socket->on<event::error>([](const Exception& e){
      std::cout << "error: " << e.message() << std::endl << std::flush;
    });

    socket->on<event::close>([](){
      std::cout << "socket closed" << std::endl << std::flush;
    });

    socket->bind(COMMON_PORT);
  });

  CHECK(received == 2);
}
